kernel/kernel-defaults/src/test/java/io/delta/kernel/defaults/benchmarks/BenchmarkParallelCheckpointReading.java

```diff
@@ -0,0 +1,249 @@
+ * benchmark                      0  avgt    5  1565.520 ±  20.551  ms/op
+ * benchmark                      1  avgt    5  1064.850 ±  19.699  ms/op
+ * benchmark                      2  avgt    5   785.918 ± 176.285  ms/op
+ * benchmark                      4  avgt    5   729.487 ±  51.470  ms/op
+ * benchmark                     10  avgt    5   693.757 ±  41.252  ms/op
+ * benchmark                     20  avgt    5   702.656 ±  19.145  ms/op
+ * }</pre>
+ */
+@State(Scope.Benchmark)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 3)
+@Fork(1)
+public class BenchmarkParallelCheckpointReading {
+
+    /**
+     * Following are the steps to generate a simple large table with multi-part checkpoint files
+     * <pre>{@code
+     * bin/spark-shell --packages io.delta:delta-spark_2.12:3.1.0 \
+     *   --conf "spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension" \
+     *   --conf "spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog" \
+     *   --conf "spark.databricks.delta.checkpoint.partSize=100000"
+     *
+     * # Within the Spark shell, run the following commands
+     * scala> spark.range(0, 100000) .withColumn("pCol", 'id % 100000) .repartition(10)
+     *   .write.format("delta") .partitionBy("pCol") .mode("append")
+     *   .save("~/test-tables/large-table")
+     *
+     * # Repeat the above steps for each of the next ranges # 100000 to 200000, 200000 to 300000 etc
+     * until enough log entries are reached.
+     *
+     * # Then create a checkpoint
+     * # This step create a multi-part checkpoint with each checkpoint containing 100K records.
+     * scala> import org.apache.spark.sql.delta.DeltaLog
+     * scala> DeltaLog.forTable(spark, "~/test-tables/large-table").checkpoint()
+     * }</pre>
+     */
+    public static final String testTablePath = "<TODO> fill the path here";
+
+    @State(Scope.Benchmark)
+    public static class BenchmarkData {
+        // Variations of number of threads to read the multi-part checkpoint files
+        // When thread count is 0, we read using the current default parquet handler implementation
+        // In all other cases we use the custom parallel parquet handler implementation defined
+        // in this benchmark
+        @Param({"0", "1", "2", "4", "10", "20"})
+        private int parallelReaderCount = 0;
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.AverageTime)
+    public void benchmark(BenchmarkData benchmarkData, Blackhole blackhole) throws Exception {
+        TableClient tableClient = createTableClient(benchmarkData.parallelReaderCount);
+        Table table = Table.forPath(
+                createTableClient(benchmarkData.parallelReaderCount),
+                testTablePath);
+
+        Snapshot snapshot = table.getLatestSnapshot(tableClient);
+        ScanBuilder scanBuilder = snapshot.getScanBuilder(tableClient);
+        Scan scan = scanBuilder.build();
+
+        Row row = scan.getScanState(tableClient);
```

Comment:
Unused?

Reply:
This is fine if it's to mimic the real case but could you add a comment explaining what we're doing in this block/the intention?

Reply:
This is just to mimic the real case. Added comments.